[jit][errors] Improve quality of error for non-implemented forward pass

bwasti opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21646 [jit][errors] Throw error when super().__init__() hasn't been called
* #21541 [jit][errors] Describe the location of a failed script attempt (to be improved when stack traces land)
* **#21540 [jit][errors] Improve quality of error for non-implemented forward pass**

Differential Revision: [D15749918](https://our.internmc.facebook.com/intern/diff/D15749918)

bwasti committed:
[jit][errors] Improve quality of error for non-implemented forward pass

driazati commented on torch/jit/__init__.py:
```diff
@@ -1696,6 +1696,9 @@ def _convert_to_script_module(mod, methods=None):
         return _create_constant_iterable_module(mod)
 
+    if mod.__class__.forward is torch.nn.Module.forward:
+        raise Exception("nn.Module must implement forward(self, ...) to be scripted")
```
This is not necessarily true, the `@torch.jit.export` decorator can be used on other functions that are acting as entry points

bwasti replied:
so forward doesn't need to be implemented?

driazati replied:
Yeah, it's not necessary

bwasti replied:
shoot.  the error for not having forward when you need it is unusable, any thoughts on how to fix it?

```
ValueError: Compiled functions can't take variable number of arguments or use keyword-only arguments with defaults
```

driazati replied:
It’s trying to compile the forward from nn.Module it looks like, we could check if nothing is exported and mod.forward is the same as nn.Module.forward and throw and error explaining what to do

bwasti replied:
`@torch.jit.export` adds the `_jit_internal.FunctionModifiers.EXPORT` annotation right?
I may be misreading this code, but it looks like
1) `_convert_to_script_module` is never called with a non-default `methods` argument
2) `methods` always includes 'forward' (lines 1702-1703 of this file)
3) we will always call `_create_methods` on 'forward'

that being said I'll move the check to instead be
```
if 'forward' in methods and mod.__class__.forward is torch.nn.Module.forward:
    # ...
```

does that work?

driazati replied:
Yeah it should be fine to just remove `forward` from the list of methods to compile if it's the same as the `torch.nn.Module.forward`, though I'm not sure if we want to error even if there's no forward and no exported methods, someone may have a `ScriptModule` they're using to just transfer some tensors to C++ or something.

bwasti committed:
Update on "[jit][errors] Improve quality of error for non-implemented forward pass"

[jit][errors] Improve quality of error for non-implemented forward pass

gh-metadata: pytorch pytorch 21540 gh/bwasti/13/head
